src/librustc_mir/interpret/eval_context.rs

```diff
@@ -23,14 +23,18 @@ use super::{
     Immediate, MPlaceTy, Machine, MemPlace, Memory, OpTy, Operand, Place, PlaceTy,
     ScalarMaybeUndef, StackPopInfo,
 };
+use rustc::infer::canonical::OriginalQueryValues;
+use rustc::infer::InferCtxt;
 
-pub struct InterpCx<'mir, 'tcx, M: Machine<'mir, 'tcx>> {
+pub struct InterpCx<'infcx, 'mir, 'tcx, M: Machine<'mir, 'tcx>> {
     /// Stores the `Machine` instance.
     pub machine: M,
 
     /// The results of the type checker, from rustc.
     pub tcx: TyCtxtAt<'tcx>,
 
+    pub(super) infcx: &'infcx InferCtxt<'infcx, 'tcx>,
```

Comment:
Yes we do, for normalizing, `Instance::resolve_xxx` calls, and recursive `const_eval` queries. We need to use this whenever we pass `param_env` into a query as it can contain inference variables.

Reply:
Why do we use inference variables? Couldn't miri be executed with the canonical form?
Canonical variables in scope would behave like type parameters, wouldn't they?

Reply:
Hm, I never thought about that, I don't think we actually need inference variables as we aren't actually resolving them. Currently it seems like the canconical form is only ever used in preparing/start of queries, so I'm not sure what problems would there be like what happens to `Bound` variables if they are part of a canonical query down the line.

It would certainly make `InferCtxt` less  invasive, which seems like a good thing. I would like to hear @nikomatsakis thought on this.

Reply:
I don't really see a problem with it.